Refactor formula.py

miskosz opened the pull request:
- Merge `Formula` and `Curve` classes into one class `Curve`
- The class stores all information to be rendered (notably min and max date to render)
- The class is initialised with a function that converts a date to a float value.

I am not totally happy with how everything is implemented but I think it is a step forward. I leave for another PR the x-axis label, which should change based on GraphType.

miskosz committed:
Refactor formulas.

miskosz committed:
Curve as a function of date.

miskosz committed:
Merge branch 'master' into refactor-formula

miskosz committed:
Fix graph offset.

miskosz committed:
Format.

miskosz commented:
This PR is still WIP:

- There is an off-by-one error for graph display, for some countries the prediction is not aligned with observation (see screenshot below)
- Log-log graphs have wrong bounds

<img width="1311" alt="Screenshot 2020-04-22 at 23 35 47" src="https://user-images.githubusercontent.com/2885431/80040457-08a1e080-84f2-11ea-865c-473cab9af51a.png">

lukipuki commented on covid_graphs/covid_graphs/formula.py:
```diff
@@ -1,81 +1,77 @@
-        ]
+        # Display values from the first day for which the number of cumulative active is
+        # more than min_case_count. This day is also "day one".
+        start_idx = np.argmax(country_report.cumulative_active >= self.min_case_count)
```
Katka and Rišo use a strict inequality, I think that's what's causing the shift for you. This is what we had before
```
return np.argmax(cumulative_active > formula.min_case_count)
```

Strict inequality is weird though, I'd suggest updating the predictions instead by bumping the minimum by +1.

miskosz replied:
Used `>=` according to the new specification. I understood that `min_cases` define `t=1` day. It looks like it works ok now, only `Austria` was adjusted manually in `predictions.py`.

lukipuki commented on covid_graphs/covid_graphs/country_graph.py:
```diff
@@ -118,30 +113,31 @@ def pick_xaxis_labels(object):
-                    x=pick_xaxis_labels(curve),
-                    y=curve.y,
-                    text=curve.date_list,
+                    x=list(map(adjust_xlabel, curve.xs)),
```
Previously, each curve was extended if it was shorter than the data of active cases. You need to take into account the length of `self.cropped_cumulative_active` when plotting it it.

lukipuki commented on covid_graphs/covid_graphs/formula.py:
```diff
@@ -1,81 +1,77 @@
-        self.maximal_idx = self.y.argmax()
+        # Display values until the second inflection point.
+        second_ip_days = math.ceil(self.tg * (self.exponent + math.sqrt(self.exponent)))
+        end_date = start_date + datetime.timedelta(days=second_ip_days)
```
This might be too short, the curve should end where the graph ends, so it needs to take into account the length of `cumulative_active` as well as other predictions plotted on the same graph.

miskosz replied:
Fixed by changing get_trace() interface.

miskosz committed:
Fix offsets. Extend x-axis to cover observations.

miskosz committed:
Fill in function comments.

miskosz commented:
Ready for review, I have mixed feelings though. Here is the changelog:

- `Formula` and `Curve` classes combined into a single `Curve` class
- Common date axis for all graphs, start_date is part of the formula
- Formulas work on dates, not "t"
- Fixed graph offsets (interpretation of `min_cases`)

Things that are not ideal:

- `Curve` remembers start and end date for a curve, which is then overridden by a wider range
- Weird caching of trace in `Curve`, just to be able to compute `maximal_y`.

miskosz commented on covid_graphs/covid_graphs/predictions.py:
```diff
@@ -25,146 +25,163 @@ class CountryPrediction:
+    #     prediction_event=OTHER,
+    #     country="Slovakia",
+    #     formula=Formula(lambda t: 8 * t ** 1.28, r"$8 \cdot t^{1.28}$", 40, 10),
+    # ),
```
Note: I have disabled Slovakia. Do we want it? I can bring it back in another PR, but AFAIK we don't display it.

lukipuki replied:
No, it's fine. I just had it there to draw possible exponents for my simulation code, it was really for convenience only.

lukipuki commented on covid_graphs/covid_graphs/country_graph.py:
```diff
@@ -28,77 +27,106 @@ class CountryGraph:
-                line=dict(width=2, dash="dash", color=color),
+
+        # Traces
+        traces = []
```
Nice restructuring!

lukipuki commented on covid_graphs/covid_graphs/formula.py:
```diff
@@ -1,81 +1,90 @@
-    """
+    def get_curve(self, country_report: CountryReport) -> Curve:
+        label = r"$\frac{_A}{_TG} \cdot \left(\frac{t}{_TG}\right)^{_expon} / e^{t/_TG}$"
+        label = (
+            label.replace("_A", f"{self.a:.0f}")
+            .replace("_TG", f"{self.tg}")
+            .replace("_expon", f"{self.exponent}")
```
This is actually a hack that I quickly made. f-strings and TeX formulas don't mix well. Is there a better way than `replace`?

miskosz replied:
Not that I know of, maybe there is :)

lukipuki commented on covid_graphs/covid_graphs/formula.py:
```diff
@@ -1,81 +1,90 @@
-        self.maximal_idx = self.y.argmax()
+        # Display values until the second inflection point.
+        second_ip_days = math.ceil(self.tg * (self.exponent + math.sqrt(self.exponent)))
+        end_date = start_date + datetime.timedelta(days=second_ip_days)
```
`end_date` is perhaps not the best name, since the curve doesn't really have an end and we might override it anyway. A possible name is `second_ip_day`, but probably not the best.

We clearly want to plot the curve at least up to the second inflection point. I'm tired and out of good ideas, but something like `display_at_least_to_date` or `min_display_end_date` would be better. I can't think of a shorter name right now.

lukipuki reviewed (approved):
Instead of caching the whole trace in `Curve`, why don't you just store `maximal_y`?

miskosz committed:
Simplify Curve.

miskosz committed:
Forgot to reformat.

miskosz commented:
About `maximal_y`, very good point :D I moved that code so many times I did not realise it can be simplified.

I have realised though that we actually **can** compute the plot range on initialisation, so I have simplified the Curve class even further.

miskosz committed:
Nits.
